userspace/libsinsp/sinsp.cpp

```diff
@@ -1306,6 +1315,78 @@ void sinsp::get_procs_cpu_from_driver(uint64_t ts)
 	}
 }
 
+int32_t sinsp::fetch_next_event(sinsp_evt*& evt)
+{
+	// check if an event must be replayed, which currently happens
+	// when a capture file is read and we discover the first "event" block
+	// after the initial "machine state" section
+	if (m_replay_scap_evt != NULL)
+	{
+		evt->m_pevt = m_replay_scap_evt;
+		evt->m_cpuid = m_replay_scap_cpuid;
+		evt->m_dump_flags = m_replay_scap_flags;
+		m_replay_scap_evt = NULL;
+		return SCAP_SUCCESS;
+	}
+
+	// start by assuming that we already have an event successfully-fetched
+	// from later that has been delayed. If our current libscap event storage
+	// is empty, attempt fetching the next event in line from the scap handle
+	int32_t res = SCAP_SUCCESS;
+	if (m_delayed_scap_evt.empty())
+	{
+		res = m_delayed_scap_evt.next(m_h);
+	}
+
+	// in case we receive a timeout (when there's no element to fetch and no
+	// error is encountered) we attempt popping an event from the asynchronous
+	// event queue. If none is available, we just return the timeout.
+	// note: the queue is optimized for checking for emptyness before popping
+	if (res == SCAP_TIMEOUT &&
+		!m_async_events_queue.empty() && m_async_events_queue.pop(m_async_evt))
+	{
+		evt = m_async_evt.get();
+		if(evt->m_pevt->ts == (uint64_t) -1)
+		{
+			evt->m_pevt->ts = get_new_ts();
+		}
+		return SCAP_SUCCESS;
+	}
+
+	// in case we successfully fetched an event, or we have one delayed from
+	// before, we check that if there is any event in the async event queue
+	// that should be returned first due to having a timestamp from earlier.
+	// the goal is to guarantee events to be fetched ordered by timestamp.
+	if(res == SCAP_SUCCESS)
+	{
+		if (!m_async_events_queue.empty())
+		{
+			const auto check_ts = [this](const sinsp_evt* evt)
+			{
+				return compare_evt_timestamps(evt->m_pevt->ts, m_delayed_scap_evt.m_pevt->ts);
+			};
+
+			// This is thread-safe as we're in a MPSC case in which
+			// sinsp::next is the single consumer
+			if (m_async_events_queue.pop_if(check_ts, m_async_evt))
+			{
+				// the async event is the one with most priority
+				evt = m_async_evt.get();
+				if(evt->m_pevt->ts == (uint64_t) -1)
```

Comment:
Just wondering, is there a way to move the `ts` initilization eg: overriding `m_async_evt.get()` or somewhere else to avoid the need to remember to make this check everytime?

Reply:
The same check is made on L1349 ;)

Reply:
I think that the very best we can do with no performance penalty is to bundle it in a macro. It's ugly to do it twice but at least it's all in the same place. I think that overriding methods would just increase the complexity here.

Reply:
I see :( ok, no problem then!